server/chat-plugins/poll.ts

```diff
@@ -259,9 +286,98 @@ export class Poll {
 		const results = this.generateResults(true);
-
 		this.room.send(`|uhtmlchange|poll${this.pollNumber}|<div class="infobox">(${this.room.tr`The poll has ended &ndash; scroll down to see the results`})</div>`);
 		this.room.add(`|html|${results}`).update();
+		this.endTimer();
+		this.room.minorActivity = null;
+		delete this.room.settings.minorActivity;
+		this.room.saveSettings();
+	}
+	save() {
+		const entry: PollData = {
+			activityId: 'poll',
+			pollNumber: this.pollNumber,
+			question: this.question,
+			supportHTML: this.supportHTML,
+			multiPoll: this.multiPoll,
+			pendingVotes: this.pendingVotes,
+			voters: this.voters,
+			voterIps: this.voterIps,
+			totalVotes: this.totalVotes,
+			timeoutMins: this.timeoutMins,
+			timerEnd: this.timerEnd,
+			isQuiz: this.isQuiz,
+			questions: [...this.options.values()],
+			queued: this.queued,
+		};
+		if (this.queued) {
+			if (!this.room.settings.minorActivityQueue) {
+				this.room.settings.minorActivityQueue = [];
+			}
+			this.room.settings.minorActivityQueue.push(entry);
+		} else {
+			this.room.settings.minorActivity = entry;
+		}
+		this.room.saveSettings();
+		return entry;
+	}
+	runTimeout(timeout: number) {
+		this.timeoutMins = timeout;
+		this.timerEnd = Date.now() + timeout * 60000;
+		this.timeout = setTimeout(() => {
+			const room = this.room;
+			if (!room) return; // do nothing if the room does not exist
+			if (room.minorActivity) room.minorActivity.end();
+			room.minorActivity = null;
+			if (room.minorActivityQueue?.length) {
+				const next = Poll.next(room);
+				if (next) {
+					room.minorActivity = room.minorActivityQueue.shift()!;
+					room.add(`|c|&|/log ${room.tr`The queued poll was started.`}`).update();
+					room.modlog({
+						action: 'POLL',
+						note: '(queued)',
+					});
+					room.minorActivity.display();
+				}
+			}
+		}, timeout * 60000);
+		this.save();
+		return this.timeout;
+	}
+	endTimer() {
+		if (!this.timeout) return;
+		clearTimeout(this.timeout);
+		this.timeoutMins = 0;
+		delete this.timerEnd;
+		return this;
+	}
+	static next(room: Room) {
+		const poll = room.minorActivityQueue?.shift();
+		if (!poll) return;
+		room.settings.minorActivityQueue!.shift();
+		if (!room.minorActivityQueue?.length) room.minorActivityQueue = null;
+		if (!room.settings.minorActivityQueue?.length) delete room.settings.minorActivityQueue;
+		room.saveSettings();
+		return poll;
+	}
+	destroy() {
+		this.room.minorActivity = null;
+	}
+}
+
+// should handle restarts and also hotpatches
+for (const room of Rooms.rooms.values()) {
+	if (room.settings.minorActivity?.activityId === 'poll') {
+		room.minorActivity?.endTimer();
+		room.minorActivity = new Poll(room, room.settings.minorActivity);
+	}
+	if (room.settings.minorActivityQueue) { // rebuild queue
+		room.minorActivityQueue = [];
+		for (const queued of room.settings.minorActivityQueue) {
+			if (queued.activityId !== 'poll') continue;
+			room.minorActivityQueue.push(new Poll(room, queued));
+		}
```

Comment:
Wait you can't put a `new Poll` into `room.settings`. `room.settings` needs to be JSON.

I don't see any benefit to this, anyway. You can and should just put a `AnnouncementData | PollData` into the `minorActivityQueue`; it'd be better for everyone, right?

Reply:
This isn't putting the Poll into room.settings, this is putting the JSON into the settings.

Reply:
This takes the JSON from the queue and rebuilds every poll in the queue.

Reply:
But, sure, i can put the JSON into the queue. That does make sense.

Reply:
Done.

Reply:
Oh, you're right, I misread. But yes, if you put JSON in the queue instead, you don't need to rebuild the queue every hotpatch/restart.